### DISCUSSION TITLE
HC-SR04 working OK but JSN-SR04T not working

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#15590] [Q&A] [answered]
Hi there!  First-time post here.

I'm using Tasmota 11.1.0 (sensor) on an ESP8266.

Connecting an HC-SR04 with a voltage divider for the echo works fine, I can measure the distance.

You can see in the scope capture, the trigger for 10uS and the response on the bottom.
![WaveFormsLiveChart](https://user-images.githubusercontent.com/7669537/167752806-deb5638c-9a04-4a94-a176-393d88308759.png)

Using the same setup, I replace the HC-SR04 with a JSN-SR04T (I need the waterproof), but I cannot get it to work.
What's weird is that Tasmota sends 4 pulse (after a restart) around 100uS with no response on the Echo pin and no other trigger after.
![WaveFormsLiveChart JSN](https://user-images.githubusercontent.com/7669537/167753236-30bf72b7-9d8b-4b3b-972e-4ba523b4d4c3.png)

I'm reading [from this page](https://protosupplies.com/product/jsn-sr04t-v3-0-waterproof-ultrasonic-range-finder/), that the JSN (setup as mode 0, like the HC-SR04) need a larger trigger of 20uS.

Has anyone got the JSN working?  The [doc here](https://tasmota.github.io/docs/HC-SR04/) says the JSN should be working like the HC?

Thanks!

**Comment by laurentppol** [comment, marked as answer]
I had no problem. Just connected, configured as HC and voila.
Some time ago. (Another Tasmota version, probably 8.5.1).
Maybe Your device is factory configured to another mode?

**Reply by SimonFili** [reply]
Checking the logs after a restart, I see:
00:00:01.778 SR4: Received ''
00:00:01.880 SR4: Received ''
00:00:01.983 SR4: Received ''
00:00:02.256 SR4: Mode 0

So the Tasmota is probably sending some pulses to see if the device is configured as serial?

From [this link](https://protosupplies.com/product/jsn-sr04t-v3-0-waterproof-ultrasonic-range-finder/), I know my device is set to mode 0, but I could try the mode 1 

> MODE 1: Automatic Serial Mode
> To enter Mode 1, the M1 pads are shorted OR a resistor value of 47K is placed across the MODE pads.
> 
> In this mode the module continuously takes measurements approximately every 100mS and outputs the distance on the TX pin at 9600 baud where it can be read using a serial RX pin on an MCU.

I did not find any documentation on different modes of operation for those devices, short of looking at the source code.

There are 5 modes total, which one is supported by Tasmota?

Any clues?

**Reply by SimonFili** [reply]
I found it!

When you short the M2 pin (mode 2) 

> MODE 2: MCU Controlled Serial Mode
> 
> To enter Mode 2, the M2 pads are shorted OR a resistor value of 120K is placed across the MODE pads.
> 
> This mode is the same as Mode 1 except that the module takes a measurement only when a trigger command of 0x55 is sent to the RX pin on the module. The module then takes a single measurement and outputs the distance on its TX pin.
> 
> The minimum trigger cycle time is 60mS

![Mode2_JSN-SR04T](https://user-images.githubusercontent.com/7669537/167837968-8b629100-0c68-4fc6-9644-64cdaada315d.png)

Tasmota output this on the log

> 00:00:01.699 SR4: Received 'FF00FDFC'
> 00:00:01.718 SR4: Received 'FF00F6F5'
> 00:00:01.738 SR4: Received 'FF00F4F3'
> 00:00:01.739 SR4: Mode 3

So mode 2 on the JSC by shorting the M2 pins, will put Tasmota in Mode 3 (serial) and works.
Log output:  "SR04":{"Distance":24.600}

The question is... where is this documented?

Thanks!

**Reply by barbudor** [reply]
> The question is... where is this documented?

It would be great to add it here : https://tasmota.github.io/docs/HC-SR04/
You can use the pencil icon on the top-right of the page to edit and submit a PR

However, by looking a little bit to the code of the driver, it looks like both mode 2 and mode 3 of the SR04T should be supported.
Here:
https://github.com/arendst/Tasmota/blob/d136c20551132558bc2850f5c13c2d59b1e0fe91/tasmota/xsns_22_sr04.ino#L167-L174
and here:
https://github.com/arendst/Tasmota/blob/d136c20551132558bc2850f5c13c2d59b1e0fe91/tasmota/xsns_22_sr04.ino#L124-L131

Mode 2 is the RECEIVER Mode
Mode 3 is the TRANSCEIVER Mode although it seems that you need to define a GPIO as "SR04 Trigger" to allow TRANSCEIVER mode
